From: Bradley Morgan <include@grrlz.net>
Subject: panic: stop CPUs that lose the panic_redirect_cpu race
Date: Tue, 7 Jul 2026 18:32:53 +0000

Loser of the redirect cmpxchg must stop, not fall through to
panic_try_start().

This issue was found by sashiko [1].

Link: https://lore.kernel.org/20260707183253.9793-1-include@grrlz.net
Link: https://sashiko.dev/#/patchset/20260705164123.18746-1-include%40grrlz.net [1]
Signed-off-by: Bradley Morgan <include@grrlz.net>
Cc: Petr Mladek <pmladek@suse.com>
Cc: Feng Tang <feng.tang@linux.alibaba.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/panic.c |    7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

--- a/kernel/panic.c~panic-stop-cpus-that-lose-the-panic_redirect_cpu-race
+++ a/kernel/panic.c
@@ -400,12 +400,9 @@ static bool panic_try_force_cpu(const ch
 	if (panic_in_progress())
 		return false;
 
-	/*
-	 * Only one CPU can do the redirect. Use atomic cmpxchg to ensure
-	 * we don't race with another CPU also trying to redirect.
-	 */
+	/* Which CPU won the race? */
 	if (!atomic_try_cmpxchg(&panic_redirect_cpu, &old_cpu, this_cpu))
-		return false;
+		return old_cpu != this_cpu;
 
 	/*
 	 * Use dynamically allocated buffer if available, otherwise
_
